packages/lodestar/src/sync/range/range.ts

```diff
@@ -0,0 +1,256 @@
+   * Add this peer to an existing chain or create a new one. The update the chains status.
+   */
+  addPeer(peerId: PeerId, localStatus: phase0.Status, peerStatus: phase0.Status): void {
+    // Compute if we should do a Finalized or Head sync with this peer
+    const syncType = getRangeSyncType(localStatus, peerStatus, this.chain);
+    this.logger.debug("Sync peer joined", {peer: peerId.toB58String(), syncType});
+
+    let startEpoch: Slot;
+    let target: ChainTarget;
+    switch (syncType) {
+      case RangeSyncType.Finalized: {
+        startEpoch = localStatus.finalizedEpoch;
+        target = {
+          slot: computeStartSlotAtEpoch(this.config, peerStatus.finalizedEpoch),
+          root: peerStatus.finalizedRoot,
+        };
+        break;
+      }
+
+      case RangeSyncType.Head: {
+        // The new peer has the same finalized (earlier filters should prevent a peer with an
+        // earlier finalized chain from reaching here).
+        startEpoch = Math.min(computeEpochAtSlot(this.config, localStatus.headSlot), peerStatus.finalizedEpoch);
+        target = {
+          slot: peerStatus.headSlot,
+          root: peerStatus.headRoot,
+        };
+        break;
+      }
+    }
+
+    // If the peer existed in any other chain, remove it.
+    // re-status'd peers can exist in multiple finalized chains, only one sync at a time
+    if (syncType === RangeSyncType.Head) {
+      this.removePeer(peerId);
+    }
+
+    this.addPeerOrCreateChain(startEpoch, target, peerId, syncType);
+    this.update(localStatus.finalizedEpoch);
+  }
+
+  /**
+   * Remove this peer from all head and finalized chains. A chain may become peer-empty and be dropped
+   */
+  removePeer(peerId: PeerId): void {
+    for (const syncChain of this.chains.values()) {
+      syncChain.removePeer(peerId);
+    }
+  }
+
+  /**
+   * Compute the current RangeSync state, not cached
+   */
+  get state(): RangeSyncState {
+    const syncingHeadTargets: ChainTarget[] = [];
+    for (const chain of this.chains.values()) {
+      if (chain.isSyncing) {
+        if (chain.syncType === RangeSyncType.Finalized) {
+          return {status: RangeSyncStatus.Finalized, target: chain.target};
+        } else {
+          syncingHeadTargets.push(chain.target);
+        }
+      }
+    }
+
+    if (syncingHeadTargets.length > 0) {
+      return {status: RangeSyncStatus.Head, targets: syncingHeadTargets};
+    } else {
+      return {status: RangeSyncStatus.Idle};
+    }
+  }
+
+  /** Full debug state for lodestar API */
+  getSyncChainsDebugState(): SyncChainDebugState[] {
+    return Array.from(this.chains.values())
+      .map((syncChain) => syncChain.getDebugState())
+      .reverse(); // Newest additions first
+  }
+
+  /** Convenience method for `SyncChain` */
+  private processChainSegment: SyncChainFns["processChainSegment"] = async (blocks) => {
+    const trusted = true; // TODO: Verify signatures
```

Comment:
On my local computer:
w/o signature verification -> 14 slots/s
w/ signature verification -> 4.5 slots/s (also seems to introduce lots of choppiness in grafana reporting)